pkg/controller/apiextensions/claim/api.go

```diff
@@ -117,3 +124,63 @@ func (a *APIBinder) Bind(ctx context.Context, cm resource.CompositeClaim, cp res
 	meta.SetExternalName(cm, meta.GetExternalName(cp))
 	return errors.Wrap(a.client.Update(ctx, cm), errUpdateClaim)
 }
+
+// An APIConnectionPropagator propagates connection details by reading
+// them from and writing them to a Kubernetes API server.
+type APIConnectionPropagator struct {
+	client resource.ClientApplicator
+	typer  runtime.ObjectTyper
+}
+
+// NewAPIConnectionPropagator returns a new APIConnectionPropagator.
+func NewAPIConnectionPropagator(c client.Client, t runtime.ObjectTyper) *APIConnectionPropagator {
+	return &APIConnectionPropagator{
+		client: resource.ClientApplicator{Client: c, Applicator: resource.NewAPIUpdatingApplicator(c)},
+		typer:  t,
+	}
+}
+
+// PropagateConnection details from the supplied resource.
+func (a *APIConnectionPropagator) PropagateConnection(ctx context.Context, to resource.LocalConnectionSecretOwner, from resource.ConnectionSecretOwner) (bool, error) {
```

Comment:
This implementation is ported from crossplane-runtime, with three differences:

* It doesn't set secret propagation annotations, because nothing uses them anymore.
* It doesn't update connection secrets if their data didn't change.
* It returns a boolean to let the caller know whether the secret was actually propagated.

Note that I've deprecated the crossplane-runtime implementation, since this was the last known caller (according to a search of the crossplane GitHub org).